tls/s2n_handshake_io.c

```diff
@@ -605,26 +635,55 @@ int s2n_conn_set_handshake_type(struct s2n_connection *conn)
     /* A handshake type has been negotiated */
     conn->handshake.handshake_type |= NEGOTIATED;
 
+    if (conn->psk_params.chosen_psk == NULL) {
+        conn->handshake.handshake_type |= FULL_HANDSHAKE;
+    }
+
     s2n_cert_auth_type client_cert_auth_type;
-    GUARD(s2n_connection_get_client_auth_type(conn, &client_cert_auth_type));
+    GUARD_AS_RESULT(s2n_connection_get_client_auth_type(conn, &client_cert_auth_type));
 
-    if (conn->mode == S2N_CLIENT && client_cert_auth_type == S2N_CERT_AUTH_REQUIRED) {
+    if (conn->mode == S2N_CLIENT && client_cert_auth_type == S2N_CERT_AUTH_REQUIRED
+            && conn->handshake.handshake_type & FULL_HANDSHAKE) {
         /* If we're a client, and Client Auth is REQUIRED, then the Client must expect the CLIENT_CERT_REQ Message */
         conn->handshake.handshake_type |= CLIENT_AUTH;
-    } else if (conn->mode == S2N_SERVER && client_cert_auth_type != S2N_CERT_AUTH_NONE) {
+    } else if (conn->mode == S2N_SERVER && client_cert_auth_type != S2N_CERT_AUTH_NONE
+            && conn->handshake.handshake_type & FULL_HANDSHAKE) {
         /* If we're a server, and Client Auth is REQUIRED or OPTIONAL, then the server must send the CLIENT_CERT_REQ Message*/
         conn->handshake.handshake_type |= CLIENT_AUTH;
     }
 
+    /* Use middlebox compatibility mode for TLS1.3 by default.
+    * For now, only disable it when QUIC support is enabled. */
+    if (!conn->config->quic_enabled) {
+        conn->handshake.handshake_type |= MIDDLEBOX_COMPAT;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+int s2n_conn_set_handshake_type(struct s2n_connection *conn)
+{
     if (IS_TLS13_HANDSHAKE(conn)) {
-        /* Use middlebox compatibility mode for TLS1.3 by default.
-         * For now, only disable it when QUIC support is enabled. */
-        if (!conn->config->quic_enabled) {
-            conn->handshake.handshake_type |= MIDDLEBOX_COMPAT;
-        }
-        conn->handshake.handshake_type |= FULL_HANDSHAKE;
+        GUARD_AS_POSIX(s2n_conn_set_tls13_handshake_type(conn));
+        return S2N_SUCCESS;
+    }
 
-        return 0;
+    S2N_ERROR_IF(conn->handshake.handshake_type & HELLO_RETRY_REQUEST, S2N_ERR_INVALID_HELLO_RETRY);
+
+    conn->handshake.handshake_type = INITIAL;
+
+    /* A handshake type has been negotiated */
+    conn->handshake.handshake_type |= NEGOTIATED;
```

Comment:
The "INITIAL" here is redundant. You just need to do conn->handshake.handshake_type = NEGOTIATED.